File: src/Blokemon.App/EconomyConfiguration.cs
using System.Globalization;
using Blokemon.Product;
using Microsoft.Extensions.Configuration;

namespace Blokemon.App;

// Both hosts resolve their economy from these settings: the server from its appsettings and
// environment, the browser build from the appsettings.json served beside its boot files.
public static class EconomyConfiguration
{
    public const string ModeKey = "Blokemon:Economy:Mode";

    public const string PackAllowanceKey = "Blokemon:Economy:PackAllowance";

    public static EconomyRules Resolve(IConfiguration configuration)
    {
        ArgumentNullException.ThrowIfNull(configuration);

        var configuredMode = configuration[ModeKey];
        if (
            string.IsNullOrWhiteSpace(configuredMode)
            || string.Equals(
                configuredMode,
                nameof(EconomyMode.Unlimited),
                StringComparison.OrdinalIgnoreCase
            )
        )
        {
            return EconomyRules.Unlimited;
        }
        if (
            !string.Equals(
                configuredMode,
                nameof(EconomyMode.ClassicScarcity),
                StringComparison.OrdinalIgnoreCase
            )
        )
        {
            throw new InvalidOperationException(
                $"{ModeKey} must be {nameof(EconomyMode.Unlimited)} or "
                    + $"{nameof(EconomyMode.ClassicScarcity)}."
            );
        }

        var configuredAllowance = configuration[PackAllowanceKey];
        var packAllowance = EconomyRules.DefaultClassicPackAllowance;
        if (!string.IsNullOrWhiteSpace(configuredAllowance))
        {
            if (
                !int.TryParse(
                    configuredAllowance,
                    NumberStyles.Integer,
                    CultureInfo.InvariantCulture,
                    out packAllowance
                )
            )
            {
                throw new InvalidOperationException($"{PackAllowanceKey} must be a whole number.");
            }
        }

        return EconomyRules
            .Classic(packAllowance)
            .Match(
                static rules => rules,
                _ =>
                    throw new InvalidOperationException(
                        $"{PackAllowanceKey} must be zero or greater."
                    )
            );
    }
}

File: src/Blokemon.App/EconomyConfiguration.fs
namespace Blokemon.App

open System
open System.Globalization
open Blokemon.Product
open Microsoft.Extensions.Configuration

// Both hosts resolve their economy from these settings: the server from its appsettings and
// environment, the browser build from the appsettings.json served beside its boot files.
module EconomyConfiguration =

    [<Literal>]
    let ModeKey = "Blokemon:Economy:Mode"

    [<Literal>]
    let PackAllowanceKey = "Blokemon:Economy:PackAllowance"

    let Resolve (configuration: IConfiguration) =
        ArgumentNullException.ThrowIfNull(configuration, nameof configuration)

        let configuredMode = configuration[ModeKey]

        if
            String.IsNullOrWhiteSpace configuredMode
            || String.Equals(
                configuredMode,
                nameof EconomyMode.Unlimited,
                StringComparison.OrdinalIgnoreCase
            )
        then
            EconomyRules.Unlimited
        else

            if
                not (
                    String.Equals(
                        configuredMode,
                        nameof EconomyMode.ClassicScarcity,
                        StringComparison.OrdinalIgnoreCase
                    )
                )
            then
                raise (
                    InvalidOperationException(
                        $"{ModeKey} must be {nameof EconomyMode.Unlimited} or "
                        + $"{nameof EconomyMode.ClassicScarcity}."
                    )
                )

            let configuredAllowance = configuration[PackAllowanceKey]

            let packAllowance =
                if String.IsNullOrWhiteSpace configuredAllowance then
                    EconomyRules.DefaultClassicPackAllowance
                else
                    match
                        Int32.TryParse(
                            configuredAllowance,
                            NumberStyles.Integer,
                            CultureInfo.InvariantCulture
                        )
                    with
                    | true, parsed -> parsed
                    | _ ->
                        raise (
                            InvalidOperationException($"{PackAllowanceKey} must be a whole number.")
                        )

            match EconomyRules.Classic packAllowance with
            | DomainResult.Succeeded rules -> rules
            | DomainResult.Failed _ ->
                raise (InvalidOperationException($"{PackAllowanceKey} must be zero or greater."))
